client/src/partials/cash/extra_payment/extra_payment.js

```diff
@@ -49,66 +46,125 @@ angular.module('bhima.controllers')
-        .success(function (rows) {
-          var items = rows.filter(function (row) {
-            return row.balance > 0;
-          });
-          items.forEach(function (row) {
-            row.debitor_first = sale.first_name;
-            row.debitor_middle = sale.middle_name;
-            row.debitor_last = sale.last_name;
-            row.transaction = row.abbr + row.reference;
-            row.cost = row.balance;
-            row.currency = 2; // values are always in $ by default
-            row.currency_id = 2; // values are always in $ by default
+    $scope.formatAccount = function (ac) {
+      if(ac){return ac.account_number + ' - ' + ac.account_txt;}
+    };
+
+    function processModels(models) {
+      angular.extend(session, models);
+      session.date = new Date();
+    }
+
+    function handleErrors(err) {
+      messenger.danger('An error occured:' + JSON.stringify(err));
+    }
+
+    function getAccount(ac) {
+      if (!ac) { return; }
+       session.configured = true;
+       session.ac = ac;
+       session.complete = true;
+    }
+
+    function search() {
+      $scope.state = 'generate';
+      session.patient = session.selected;
+      session.configured = true;
+
+      var id = session.patient.debitor_uuid,
+        account_id = session.patient.account_id;
+
+      dependencies.sales = {
+        query : {
+          tables : {
+            'sale' : {
+              columns : ['uuid', 'debitor_uuid']
+            },
+            'patient' : {
+              columns : ['first_name', 'middle_name', 'last_name']
+            }
+          },
+          join : ['sale.debitor_uuid=patient.debitor_uuid'],
+          where : ['patient.debitor_uuid=' + id]
+        }
+      };
+      validate.process(dependencies, ['sales'])
+      .then(function (model) {
+        var sales = model.sales.data;
+        $scope.saleData = [];
+        sales.forEach(function (sale) {
+          $http.get('/ledgers/debitor/'+sale.debitor_uuid)
+          .success(function (rows) {
+            var items = rows.filter(function (row) {
+              return row.balance > 0;
+            });
+            items.forEach(function (row) {
+              row.debitor_first = sale.first_name;
+              row.debitor_middle = sale.middle_name;
+              row.debitor_last = sale.last_name;
+              row.transaction = row.abbr + row.reference;
+              row.cost = row.balance;
+              row.currency = $scope.project.currency_id;
+              row.currency_id = $scope.project.currency_id;
+            });
+            $scope.saleData = $scope.saleData.concat(items);
+            $scope.saleDataUnique = filterUnique($scope.saleData, 'transaction');
           });
-          $scope.saleData = $scope.saleData.concat(items);
-          $scope.saleDataUnique = filterUnique($scope.saleData, 'transaction');
-        });
+        });  
       });
+
+      function filterUnique(array, key) {
```

Comment:
Okay, so this case is more complex.  Here is a combined map/reduce that will perform what @mbayopanda was trying to do:
```
// filters and array of objects, returning a unique key/value
var uniqueValues = array.map(function (row) {
  return row[key];
}).reduce(function (aggregate, row) {
  if (aggregate.indexOf(item) === -1) { aggregate.push(item); }
  return aggregate;
}, []);
```